### PR TITLE
 [safepoint] Use LLVM pass to insert safepoint 

### PR DESCRIPTION
**Pull request by luhenry**
Depends on https://github.com/mono/llvm/pull/35

### CODE REVIEW

File: mono/mini/mini-llvm.c [hunk 1]
```diff
@@ -5937,6 +5976,10 @@ process_bb (EmitContext *ctx, MonoBasicBlock *bb)
 			break;
 		}
 		case OP_GC_SAFE_POINT: {
+			/* LLVM pass PlaceSafepoints will take care of that. See LLVMSetGC */
+			if (cfg->compile_aot && !cfg->llvm_only)
```
Comment (alexanderkyte) [inline mono/mini/mini-llvm.c]: Why is this disabled for llvmonly? The bitcode we generate should be the output of any passes we add to `mono/llvm`. This isn't in the actual codegen/Xcode-clang step is it? 

Is this optimization going to be enabled on wasm? We're really sensitive to code size there.

Comment (luhenry) [reply mono/mini/mini-llvm.c]: Because this relies on LLVM having a `MonoGC` which certainely doesn't exist in Apple's LLVM on their end.

Comment (vargaz) [reply mono/mini/mini-llvm.c]: Wouldn't this lead to a perf loss ? The current code checks sreg1 and only makes the call if its set. The new code will call gc.safepoint_poll which will a c function in  the runtime.

File: mono/mini/mini.c [hunk 2]
```diff
@@ -2953,16 +2940,16 @@ mono_insert_safepoints (MonoCompile *cfg)
 	 * we can skip the GC safepoint on method entry. */
 	gboolean requires_safepoint = cfg->has_calls;
 
+	if (requires_safepoint)
+		insert_safepoint (cfg, cfg->bb_entry);
```
Comment (lewurm) [inline mono/mini/mini.c]: why was this moved? This should be after the loop (it potentially changes `requires_safepoint`).

File: mono/mini/mini-llvm.c (hunk 1 above)
Comment (luhenry) [reply mono/mini/mini-llvm.c]: @vargaz this is indeed a problem right now, and I won't want this PR to be merged until that's fixed. I just need to figure out my way with emitting the right LLVM IR.

Comment (luhenry) [reply mono/mini/mini-llvm.c]: @vargaz that is now fixed.

File: mono/mini/mini-llvm.c [hunk 3]
```diff
@@ -7336,6 +7401,8 @@ emit_method_inner (EmitContext *ctx)
 
 	if (!cfg->llvm_only)
 		LLVMSetFunctionCallConv (method, LLVMMono1CallConv);
+	if (!cfg->llvm_only && cfg->compile_aot && mono_threads_are_safepoints_enabled ())
```
Comment (lambdageek) [inline mono/mini/mini-llvm.c]: After #11833 , call `mono_threads_target_are_safepoints_enabled ()`

File: mono/mini/mini.c [hunk 4]
```diff
@@ -2893,19 +2889,28 @@ Those are:
...
 */
 static void
-mono_insert_safepoints (MonoCompile *cfg)
+insert_safepoints (MonoCompile *cfg)
 {
 	MonoBasicBlock *bb;
 
-	if (!mono_threads_are_safepoints_enabled ())
-		return;
+	g_assert (mono_threads_are_safepoints_enabled ());
```
Comment (lambdageek) [inline mono/mini/mini.c]: Same - after #11833 lands call `mono_threads_target_are_safepoints_enabled ()`.

File: mono/mini/mini.c [hunk 5]
```diff
@@ -3636,8 +3628,10 @@ mini_method_compile (MonoMethod *method, guint32 opts, MonoDomain *domain, JitFl
 		MONO_TIME_TRACK (mono_jit_stats.jit_compute_natural_loops, mono_compute_natural_loops (cfg));
 	}
 
-	MONO_TIME_TRACK (mono_jit_stats.jit_insert_safepoints, mono_insert_safepoints (cfg));
-	mono_cfg_dump_ir (cfg, "insert_safepoints");
+	if (mono_threads_are_safepoints_enabled ()) {
```
Comment (lambdageek) [inline mono/mini/mini.c]: Andone more. `mono_threads_target_are_safepoints_enabled()` if #11833 lands first.

**Comment by lewurm** [comment]
To continue the example from @EgorBo from https://github.com/mono/mono/pull/11729:

```csharp
using System;

public class Hello {
        public static void Main (string []args)
        {
                System.Console.WriteLine ("sum: " + SumUp());
        }

        public static int SumUp ()
        {
                int sum = 0;
                for (int i = 0; i < 5; i++)
                        sum += i;
                return sum;
        }
}
```

Before:
```asm
_Hello_SumUp:
     960:   53  pushq   %rbx
     961:   48 8b 1d 68 09 00 00    movq    2408(%rip), %rbx
     968:   48 83 3b 00     cmpq    $0, (%rbx)
     96c:   74 05   je  5 <_Hello_SumUp+0x13>
     96e:   e8 dd 00 00 00  callq   221 <plt__jit_icall_mono_threads_state_poll>
     973:   48 83 3b 00     cmpq    $0, (%rbx)
     977:   74 05   je  5 <_Hello_SumUp+0x1e>
     979:   e8 d2 00 00 00  callq   210 <plt__jit_icall_mono_threads_state_poll>
     97e:   48 83 3b 00     cmpq    $0, (%rbx)
     982:   74 05   je  5 <_Hello_SumUp+0x29>
     984:   e8 c7 00 00 00  callq   199 <plt__jit_icall_mono_threads_state_poll>
     989:   48 83 3b 00     cmpq    $0, (%rbx)
     98d:   74 05   je  5 <_Hello_SumUp+0x34>
     98f:   e8 bc 00 00 00  callq   188 <plt__jit_icall_mono_threads_state_poll>
     994:   48 83 3b 00     cmpq    $0, (%rbx)
     998:   74 05   je  5 <_Hello_SumUp+0x3f>
     99a:   e8 b1 00 00 00  callq   177 <plt__jit_icall_mono_threads_state_poll>
     99f:   48 83 3b 00     cmpq    $0, (%rbx)
     9a3:   74 05   je  5 <_Hello_SumUp+0x4a>
     9a5:   e8 a6 00 00 00  callq   166 <plt__jit_icall_mono_threads_state_poll>
     9aa:   b8 0a 00 00 00  movl    $10, %eax
     9af:   5b  popq    %rbx
     9b0:   c3  retq
```
LLVM unrolls the loop and can do constant folding, but doesn't understand the semantics of our safepoints and thus can't squash them together.

With this PR:
```asm
_Hello_SumUp:
     9c0:   48 83 3d 08 09 00 00 00     cmpq    $0, 2312(%rip)
     9c8:   74 0b   je  11 <_Hello_SumUp+0x15>
     9ca:   50  pushq   %rax
     9cb:   ff 15 07 09 00 00   callq   *2311(%rip)
     9d1:   48 83 c4 08     addq    $8, %rsp
     9d5:   b8 0a 00 00 00  movl    $10, %eax
     9da:   c3  retq
```
LLVM can be awesome. Nice work @luhenry  👍

**Comment by EgorBo** [comment]
Just a random question,
I have a method (from the `dotnet/performance` repo actually):
```csharp
[MethodImpl(MethodImplOptions.NoInlining)]
public static unsafe bool LoopGoto(String strA, String strB)
{
    int length = strA.Length;

    fixed (char* ap = strA) fixed (char* bp = strB)
    {
        char* a = ap;
        char* b = bp;

        while (length != 0)
        {
            int charA = *a;
            int charB = *b;

            if (charA != charB)
                goto ReturnFalse;  // placement of prolog for this return is the issue

            a++;
            b++;
            length--;
        }

        return true;

        ReturnFalse:
        return false;
    }
}
```
mono-master LLVM-aot output: https://gist.github.com/EgorBo/c0141b5cec4749e859a5e69ececd563f
luhenry:llvm-safepoints LLVM-aot output: https://gist.github.com/EgorBo/e07dc91ea42db8a423291f9f79e8099c
Does anything look suspecious? this benchmark is 20% slower in this branch.

both branches were compiled using the same parameters

**Comment by EgorBo** [comment]
C#:
```
public static unsafe void Zeroing(int* n, int len)
{
    for (int i =0; i<len; i++)
        n[i] = 0;
}
```

LLVM IR:
```llvm
; Function Attrs: norecurse nounwind readonly uwtable
define hidden monocc i32 @P_GetValueFoo_int_(i64* nocapture readonly %arg_n) #3 gc "mono" {
BB0:
  %0 = bitcast i64* %arg_n to i32*
  %t20 = load i32, i32* %0, align 4
  %1 = load i64, i64* bitcast (i64** getelementptr inbounds ([27 x i64*], [27 x i64*]* @mono_aot_Program_llvm_got, i64 0, i64 24) to i64*), align 16
  %2 = icmp eq i64 %1, 0
  br i1 %2, label %gc.safepoint_poll.exit, label %gc.safepoint_poll.poll.i

gc.safepoint_poll.poll.i:                         ; preds = %BB0
  %3 = load void ()*, void ()** bitcast (i64** getelementptr inbounds ([27 x i64*], [27 x i64*]* @mono_aot_Program_llvm_got, i64 0, i64 25) to void ()**), align 8
  call void %3() #8
  br label %gc.safepoint_poll.exit

gc.safepoint_poll.exit:                           ; preds = %BB0, %gc.safepoint_poll.poll.i
  ret i32 %t20
}
```

^ temp.opt.bc (`mono --aot=llvm,llvmopts="-S -O0" Program.cs`)

Is it possible to figure out that all variables are value types and safe-points are not necessary here?
(I am trying to create conditions for LLVM to be able to vectorize my loops with SSE/AVX)

One more weird example:

```csharp
public static unsafe int GetValue(int* n)
{
    return *n;
}
```
Translates to:
```llvm
define hidden monocc i32 @P_GetValue_int_(i64* nocapture readonly %arg_n) #3 gc "mono" {
BB0:
  %0 = bitcast i64* %arg_n to i32*
  %t20 = load i32, i32* %0, align 4
  %1 = load i64, i64* bitcast (i64** getelementptr inbounds ([27 x i64*], [27 x i64*]* @mono_aot_Program_llvm_got, i64 0, i64 24) to i64*), align 16
  %2 = icmp eq i64 %1, 0
  br i1 %2, label %gc.safepoint_poll.exit, label %gc.safepoint_poll.poll.i

gc.safepoint_poll.poll.i:                         ; preds = %BB0
  %3 = load void ()*, void ()** bitcast (i64** getelementptr inbounds ([27 x i64*], [27 x i64*]* @mono_aot_Program_llvm_got, i64 0, i64 25) to void ()**), align 8
  call void %3() #8
  br label %gc.safepoint_poll.exit

gc.safepoint_poll.exit:                           ; preds = %BB0, %gc.safepoint_poll.poll.i
  ret i32 %t20
}
```

Could be just:
```llvm
define hidden monocc i32 @P_GetValue_int_(i32* nocapture readonly %arg_n) #3 gc "mono" {
  %0 = load i32, i32* %arg_n, align 4
  ret i32 %0
}
```
(i32* instead of i64*+bitcast for my `int* n` and no safepoints)

**Comment by luhenry** [comment]
@EgorBo that's because the LLVM backend bypasses the optimisation we have in `insert_safepoints` where we discard the leaf methods.

@lewurm any guidance on where we should do it?

**Comment by lambdageek** [comment]
Do we still want this in its current form, @luhenry ?  Looks like there's some merge conflict to resolve

**Comment by luhenry** [comment]
@lambdageek if you want to take it over, that would be most useful. The latest issue is with finding the right linking strategy for `gc.safepoint_poll` to work for fullAOT.

**Comment by lambdageek** [comment]
> The latest issue is with finding the right linking strategy for gc.safepoint_poll to work for fullAOT.

@luhenry you mean we have to decide which AOT image gets the safepoint function?

wouldn't we want it to have internal (static) linkage and an inline hint so that every AOT image has its own copy that is inlined at the call sites?

**Comment by luhenry** [comment]
I do not remember fully what was the issue it was running into. What I remember is there would be either conflicts on the symbol `gc.safepoint_poll` or the `-place-safepoints` pass would not find the symbol.

**Comment by luhenry** [comment]
@monojenkins build

**Comment by alexanderkyte** [comment]
so the crash happens because we jump to an offset in the GOT that is null. 

https://gist.github.com/alexanderkyte/634c9f1f25b7eb252e4d5fb4ea3c2fcf

**Comment by vargaz** [comment]
Not sure why that happens, I'd suggest adding it to
preinited_jit_icalls in aot-compiler.c so its initialized on startup.

**Comment by alexanderkyte** [comment]
Thanks @vargaz, trying that now.

**Comment by alexanderkyte** [comment]
```
ibrary -out:../../class/lib/net_4_x-linux/System.Web.Http.WebHost.dll  @./../../build/deps/linux_net_4_x__System.Web.Http.WebHost.dll.response
Read in 33 resources from '/mnt/jenkins/workspace/test-mono-pull-request-i386/external/aspnetwebstack/src/System.Web.Http.SelfHost/Properties/SRResources.resx'
Writing resource file...  Done.
MONO_PATH="./../../class/lib/build:$MONO_PATH" CSC_SDK_PATH_DISABLED= /mnt/jenkins/workspace/test-mono-pull-request-i386/runtime/mono-wrapper  --aot-path=/mnt/jenkins/workspace/test-mono-pull-request-i386/mcs/class/lib/build --gc-params=nursery-size=64m /mnt/jenkins/workspace/test-mono-pull-request-i386/external/roslyn-binaries/Microsoft.Net.Compilers/Microsoft.Net.Compilers.2.8.2/tools/csc.exe  /codepage:65001 /nologo /noconfig /deterministic   -d:NET_4_0 -d:NET_4_5 -d:NET_4_6 -d:MONO -d:WIN_PLATFORM -nowarn:1699 -nostdlib /debug:portable -optimize /features:peverify-compat /langversion:latest  -d:ASPNETMVC -keyfile:../winfx.pub -delaysign -r:./../../class/lib/net_4_x-linux/System.Core.dll -r:./../../class/lib/net_4_x-linux/System.dll -r:./../../class/lib/net_4_x-linux/System.Xml.dll -r:./../../class/lib/net_4_x-linux/System.Configuration.dll -r:./../../class/lib/net_4_x-linux/System.Net.Http.dll -r:./../../class/lib/net_4_x-linux/System.Runtime.Serialization.dll -r:./../../class/lib/net_4_x-linux/System.ServiceModel.dll -r:./../../class/lib/net_4_x-linux/System.IdentityModel.dll -r:./../../class/lib/net_4_x-linux/System.Web.Http.dll -r:./../../class/lib/net_4_x-linux/System.Net.Http.Formatting.dll -r:./../../class/lib/net_4_x-linux/mscorlib.dll  -sourcelink:./../../build/common/sourcelink.json  -resource:System.Web.Http.SelfHost.Properties.CommonWebApiResources.resources -resource:System.Web.Http.SelfHost.Properties.SRResources.resources -target:library -out:../../class/lib/net_4_x-linux/System.Web.Http.SelfHost.dll  @./../../build/deps/linux_net_4_x__System.Web.Http.SelfHost.dll.response

Unhandled Exception:
StackOverflowException
[ERROR] FATAL UNHANDLED EXCEPTION: System.StackOverflowException: The requested operation caused a stack overflow.
  at (wrapper managed-to-native) System.Object.__icall_wrapper_mono_thread_interruption_checkpoint()
  at (wrapper managed-to-native) System.Object.__icall_wrapper_mono_threads_state_poll()
  at (wrapper managed-to-native) System.Object.__icall_wrapper_mono_thread_interruption_checkpoint()
  at (wrapper managed-to-native) System.Object.__icall_wrapper_mono_threads_state_poll()
  at (wrapper managed-to-native) System.Object.__icall_wrapper_mono_thread_interruption_checkpoint()
  at (wrapper managed-to-native) System.Object.__icall_wrapper_mono_threads_state_poll()
  at (wrapper managed-to-native) System.Object.__icall_wrapper_mono_thread_interruption_checkpoint()
  at (wrapper managed-to-native) System.Object.__icall_wrapper_mono_threads_state_poll()
  at (wrapper managed-to-native) System.Object.__icall_wrapper_mono_thread_interruption_checkpoint()
  at (wrapper managed-to-native) System.Object.__icall_wrapper_mono_threads_state_poll()
  at (wrapper managed-to-native) System.Object.__icall_wrapper_mono_thread_interruption_checkpoint()
  at (wrapper managed-to-native) System.Object.__icall_wrapper_mono_threads_state_poll()
  at (wrapper managed-to-native) System.Object.__icall_wrapper_mono_thread_interruption_checkpoint()
  at (wrapper managed-to-native) System.Object.__icall_wrapper_mono_threads_state_poll()
  at (wrapper managed-to-native) System.Object.__icall_wrapper_mono_thread_interruption_checkpoint()
```

Will fix

**Comment by alexanderkyte** [comment]
@lambdageek seems a bit weird to be seeing this on the OSX desktop lane. It seems like it's definitely related to the area of the runtime that this changes though.

File: mono/mini/mini.c [hunk 6]
```diff
@@ -2872,6 +2872,19 @@ insert_safepoints (MonoCompile *cfg)
 		return;
 	}
 
+	if (cfg->method->wrapper_type == MONO_WRAPPER_MANAGED_TO_NATIVE) {
```
Comment (luhenry) [inline mono/mini/mini.c]: Just merge that case with the one above at https://github.com/mono/mono/pull/11789/files/4200cc62bff4c880713c335a2de8f9135e7092ae..b06eb38ee0a4bfefdfdab0911f438a396c453d5f#diff-dbbf26d27078d673bee0d8d752e3b903R2858. 
```
if (info && info->subtype == WRAPPER_SUBTYPE_ICALL_WRAPPER &&
	(info->d.icall.func == poll_func ||
	 info->d.icall.func == mono_thread_interruption_checkpoint ||
 	 info->d.icall.func == mono_threads_exit_gc_safe_region_unbalanced)) { ... }
```

Comment (luhenry) [reply mono/mini/mini.c]: And that's funny, that's how the code was before https://github.com/mono/mono/pull/11789/commits/4200cc62bff4c880713c335a2de8f9135e7092ae.

**Comment by alexanderkyte** [comment]
@monojenkins build failed

**Comment by alexanderkyte** [comment]
Seeing some failures on `MonoTests.runtime.bug-60862.exe`, which generates a stacktrace. We mis-report it as being a stacktrace in unmanaged. That seems like it would happen if we do a stacktrace while we are at the top of the stack. 

Cannot reproduce on OSX at all. Will attempt at home on linux machine. 

@lambdageek would a safepoint triggering on an overflowing stack be correctly identified? I'm not sure what of the many types of fixes for that would be the least disruptive.

**Comment by lambdageek** [comment]
> would a safepoint triggering on an overflowing stack be correctly identified? I'm not sure what of the many types of fixes for that would be the least disruptive.

If you question is "can the code for a safepoint be called when the stack is overflowing"? I'm pretty sure the answer is no:
1. it'll behave like any other pure C code stack overflow.
2. if you call the safepoint code from the sigaltstack, (and a safepoint is needed and performed), it will be confused - the stack copying code expects a stack pointer to be in bounds for the thread.

If I had to guess, LLVM and mono's old mechanism are placing the safepoint in slightly different places with respect to the `try`/`catch` and that's somehow resulting in a stack overflow in the safepoint code instead of earlier in managed code.

**Comment by alexanderkyte** [comment]
Difficulties reproducing on my linux desktop, but I think I have a good defensive strategy after looking over it. Will push a fix tomorrowish and see what CI finds.

**Comment by alexanderkyte** [comment]
After some poking, I don't think I'm going to be able to prevent this without reproducing. Let me redouble my efforts and try an ubuntu VM on my linux host with the same version we use.

**Comment by alexanderkyte** [comment]
I've been able to repro now. Looking into it.

**Comment by alexanderkyte** [comment]
Looks like it's trying to lazily init the trampoline during the StackOverflow.

```
  thread #4, name = 'tid_2603'
    frame #0: 0x00007fff6c0dbd82 libsystem_kernel.dylib`__semwait_signal + 10
    frame #1: 0x00007fff6c056724 libsystem_c.dylib`nanosleep + 199
    frame #2: 0x000000010f3c82a3 mono`monoeg_g_usleep(microseconds=1000000) at gdate-unix.c:54
    frame #3: 0x000000010ef616d9 mono`mono_handle_soft_stack_ovf(jit_tls=0x00007f9485043600, ji=0x0000000000000000, ctx=0x0000000110dc6f48, siginfo=0x0000000110dc6ee0, fault_addr="") at mini-exceptions.c:3174
    frame #4: 0x000000010ee572fe mono`mono_sigsegv_signal_handler_debug(_dummy=10, _info=0x0000000110dc6ee0, context=0x0000000110dc6f48, debug_fault_addr=0x0000700005ac5ff8) at mini-runtime.c:3365
    frame #5: 0x000000010ee57133 mono`mono_sigsegv_signal_handler(_dummy=10, _info=0x0000000110dc6ee0, context=0x0000000110dc6f48) at mini-runtime.c:3434
    frame #6: 0x00007fff6c299f5a libsystem_platform.dylib`_sigtramp + 26
    frame #7: 0x000000010f39e7e5 mono`mono_memory_write_barrier at mono-membar.h:77
    frame #8: 0x000000010f39eafb mono`mono_get_hazardous_pointer(pp=0x00007f9483f38a28, hp=0x0000000110cae648, hazard_index=1) at hazard-pointer.c:214
    frame #9: 0x000000010f1cee9e mono`jit_info_table_chunk_index(chunk=0x00007f9483f38950, hp=0x0000000110cae648, addr=0x0000000110d1c120) at jit-info.c:200
    frame #10: 0x000000010f1cdb0a mono`jit_info_table_find(table=0x00007f9483f38680, hp=0x0000000110cae648, addr=0x0000000110d1c120) at jit-info.c:222
    frame #11: 0x000000010f1cd8f6 mono`mono_jit_info_table_find_internal(domain=0x00007f9483f11e80, addr=0x0000000110d1c120, try_aot=1, allow_trampolines=0) at jit-info.c:293
    frame #12: 0x000000010ef5c78b mono`mini_jit_info_table_find_ext(domain=0x00007f9483f11e80, addr=0x0000000110d1c120, allow_trampolines=0, out_domain=0x0000000000000000) at mini-exceptions.c:1936
    frame #13: 0x000000010ef5e9a9 mono`mini_jit_info_table_find(domain=0x00007f9483f11e80, addr=0x0000000110d1c120, out_domain=0x0000000000000000) at mini-exceptions.c:1974
    frame #14: 0x000000010ef67008 mono`mini_add_method_trampoline(m=0x00007f948600c290, compiled_method=0x0000000110d1c120, add_static_rgctx_tramp=0, add_unbox_tramp=0) at mini-trampolines.c:338
    frame #15: 0x000000010ef6822a mono`common_call_trampoline(regs=0x0000700005ac6598, code="?????\x88D$\x06\x8aD$\x06\x88D$\a\x8aD$\aH\x83\U00000098\U00000088\x89D$\x10H\x8bD$\x10?", m=0x00007f948600c290, vt=0x0000000000000000, vtable_slot=0x0000000000000000, error=0x0000700005ac64c8) at mini-trampolines.c:643
    frame #16: 0x000000010ef673f8 mono`mono_magic_trampoline(regs=0x0000700005ac6598, code="?????\x88D$\x06\x8aD$\x06\x88D$\a\x8aD$\aH\x83\U00000098\U00000088\x89D$\x10H\x8bD$\x10?", arg=0x00007f948600c290, tramp="???") at mini-trampolines.c:771
    frame #17: 0x0000000110c4d393
    frame #18: 0x0000000110dbc679 bug-60862.exe.dylib`StackOverflowTest_Program_WaitOne_bool_bool + 25
    frame #19: 0x0000000110dbc67e bug-60862.exe.dylib`StackOverflowTest_Program_WaitOne_bool_bool + 30
    frame #20: 0x0000000110dbc67e bug-60862.exe.dylib`StackOverflowTest_Program_WaitOne_bool_bool + 30
```

**Comment by alexanderkyte** [comment]
Looks like it will break outside of init as well. Why this wasn't happening before, I'm not sure. 

```
    frame #0: 0x00007fff6c0dbd82 libsystem_kernel.dylib`__semwait_signal + 10
    frame #1: 0x00007fff6c056724 libsystem_c.dylib`nanosleep + 199
    frame #2: 0x000000010d6fc2a3 mono`monoeg_g_usleep(microseconds=1000000) at gdate-unix.c:54
    frame #3: 0x000000010d2956d9 mono`mono_handle_soft_stack_ovf(jit_tls=0x00007f9e99832e00, ji=0x0000000000000000, ctx=0x000000010f178f48, siginfo=0x000000010f178ee0, fault_addr="") at mini-exceptions.c:3174
    frame #4: 0x000000010d18b2fe mono`mono_sigsegv_signal_handler_debug(_dummy=10, _info=0x000000010f178ee0, context=0x000000010f178f48, debug_fault_addr=0x0000700002ce3ffc) at mini-runtime.c:3365
    frame #5: 0x000000010d18b133 mono`mono_sigsegv_signal_handler(_dummy=10, _info=0x000000010f178ee0, context=0x000000010f178f48) at mini-runtime.c:3434
    frame #6: 0x00007fff6c299f5a libsystem_platform.dylib`_sigtramp + 26
    frame #7: 0x000000010d6e0c00 mono`mono_threads_suspend_policy at mono-threads-coop.c:583
    frame #8: 0x000000010d6e0159 mono`mono_threads_is_blocking_transition_enabled at mono-threads-coop.h:81
    frame #9: 0x000000010d6dfef1 mono`mono_threads_state_poll_with_info(info=0x00007f9e99831800) at mono-threads-coop.c:123
  * frame #10: 0x000000010d6dfed1 mono`mono_threads_state_poll at mono-threads-coop.c:117
    frame #11: 0x00000001113b73d1 mscorlib.dll.dylib`wrapper_managed_to_native_object___icall_wrapper_mono_threads_state_poll + 81
    frame #12: 0x000000010f16e679 bug-60862.exe.dylib`StackOverflowTest_Program_WaitOne_bool_bool + 25
    frame #13: 0x000000010f16e67e bug-60862.exe.dylib`StackOverflowTest_Program_WaitOne_bool_bool + 30
    frame #14: 0x000000010f16e67e bug-60862.exe.dylib`StackOverflowTest_Program_WaitOne_bool_bool + 30
    frame #15: 0x000000010f16e67e bug-60862.exe.dylib`StackOverflowTest_Program_WaitOne_bool_bool + 30
    frame #16: 0x000000010f16e67e bug-60862.exe.dylib`StackOverflowTest_Program_WaitOne_bool_bool + 30
```

**Comment by luhenry** [comment]
We are still observing the following error on iOS SDK:
```
duplicate symbol _gc.safepoint_poll in:
    /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/aot-cache/mscorlib.dll.llvm.o
    /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/aot-cache/I18N.dll.llvm.o
duplicate symbol _gc.safepoint_poll in:
    [...]
```

**Comment by alexanderkyte** [comment]
@monojenkins build failed

**Comment by alexanderkyte** [comment]
@luhenry interesting, I hadn't been looking for that one. Looks like a static linking issue. Will fix.

**Comment by alexanderkyte** [comment]
Looks like this should be good to merge once those two arm lanes come back.

**Comment by luhenry** [comment]
@alexanderkyte thanks for pushing it through! :)

**Pull request merged**